Kernel/Net/IP/Socket.cpp

```diff
@@ -108,74 +159,140 @@ ErrorOr<void> IPv4Socket::ensure_bound()
     return result;
 }
 
-ErrorOr<void> IPv4Socket::bind(Credentials const& credentials, Userspace<sockaddr const*> user_address, socklen_t address_size)
+// For the generic port retrieval code to work (see below in bind()), port fields need to be identical between IP versions.
+static_assert(__builtin_offsetof(sockaddr_in, sin_port) == __builtin_offsetof(sockaddr_in6, sin6_port)
+    && sizeof(decltype(declval<sockaddr_in>().sin_port)) == sizeof(decltype(declval<sockaddr_in6>().sin6_port)));
+
+static ErrorOr<in_port_t> copy_port_from_user(Userspace<sockaddr const*> user_address)
+{
+    in_port_t requested_local_port_network_order;
+    auto* address = reinterpret_cast<sockaddr_in const*>(user_address.unsafe_userspace_ptr());
+    // Since the port is in the same place for IPv4 and IPv6, this will always be correct.
+    TRY(copy_from_user(&requested_local_port_network_order, &address->sin_port, sizeof(in_port_t)));
+    auto requested_local_port = AK::convert_between_host_and_network_endian(requested_local_port_network_order);
+    return requested_local_port;
+}
```

Comment:
Passerby comment: this feels overly complicated for what it achieves. Maybe add an additional method in the delegate that extracts port or just make `set_local_address` return port since you're reading sockaddr there already?
(Feel free to ignore)